### DISCUSSION TITLE
Postgresql views

### DISCUSSION THREAD

**Discussion by micebrain** [Budibase/budibase#4111] [Help] [answered]
Hi,

I added a view in my Postgresql DB. How do I make it available in the data source? When I click on "Fetch table" I get the following error:

```
Error fetching tables
No primary key constraint found for the following: view_item_owner
```

Thanks!

**Comment by mjashanks** [comment, marked as answer]
Ah yes - when you use the "Fetch Tables" feature, it only returns tables that have a primary key. It needs a primary key to be able to Update and Delete. 

If you want to use views, you have to use a custom query.

In future, it'd be good to list views as well (or tables without a PK), but disable Update/Delete queries.

**Reply by micebrain** [reply]
Thanks for the explanation. Yes, the intention is only for display purposes and I was thinking that using view BB will treat it like a table and I can get the pagination for free - since pagination does not work with query.

**Reply by marco-santulli** [reply]
I bumped into the same problem today.

**Reply by sdsys-ch** [reply]
I've run into this as well in the past and "solved" it via materialized views. However, the impact of refreshing these via trigger(function)s upon INSERTS/etc on the parent tables is getting a little too much with large datasets.

It would be great if we could treat views / tables without a PKEY as described by @mjashanks with READ only.
+1

**Reply by mmorley0395** [reply]
this indeed would be a great feature- would rather have the flexibility of pre-creating some views for read-only purposes before loading into budibase

**Reply by mmorley0395** [reply]
> I've run into this as well in the past and "solved" it via materialized views. However, the impact of refreshing these via trigger(function)s upon INSERTS/etc on the parent tables is getting a little too much with large datasets.
> 
> It would be great if we could treat views / tables without a PKEY as described by @mjashanks with READ only. +1

would you mind sharing how you 'solved' this even in the interim? i can't seem to get budibase to recognize/find materialized views when fetching
